validation: Add Origin field to field.Error for more precise error tracking

yongruilin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:
This PR introduces a new `Origin` field to the Error type and updates validation tests to use it, improving test robustness and readability.
Main change: 
- Added Origin field to Error and related methods for better error comparison in tests.

This PR also converts 2 places as example to demonstrate the usage of `Origin`:
- `ValidateEndpointsCreate()`
- Validation of `ReplicationController.spec.Replicas`

To mitigate the collateral damage:
- Replaced manual error logging with cmp.Diff and cmpopts in resource validation tests.

These changes make validation tests less brittle by reducing reliance on detailed error strings, improving maintainability and debugging clarity.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #
Part of https://github.com/kubernetes/enhancements/issues/5073
#### Special notes for your reviewer:
- Recommend to review by commits.
- It inherits from https://github.com/kubernetes/kubernetes/pull/130055

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md
```

yongruilin committed:
feat: Add Origin field to Error and related methods

This change introducing a new field in Error. It would be used in testing to compare the expected errors without matching the detail strings.

Co-authored-by: Tim Hockin <thockin@google.com>

yongruilin committed:
test: Improve error comparison in resource validation tests

Replace manual error logging with cmp.Diff for more precise error comparisons, using cmpopts to ignore Origin field and support UniqueString comparison.

yongruilin committed:
test: convert ValidateEndpointsCreate to use error Origin field in test

Update ValidateEndpointsCreate validation tests to use the new Origin field for more precise error comparisons. It leverage the Origin field instead of detailed error messages, improving test robustness and readability.

Co-authored-by: Tim Hockin <thockin@google.com>

yongruilin committed:
test: Add Origin field support to ReplicationController spec.Replicas validation test

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6266,7 +6266,7 @@ func ValidateReplicationControllerSpec(spec, oldSpec *core.ReplicationController
 	allErrs = append(allErrs, ValidateNonnegativeField(int64(spec.MinReadySeconds), fldPath.Child("minReadySeconds"))...)
 	allErrs = append(allErrs, ValidateNonEmptySelector(spec.Selector, fldPath.Child("selector"))...)
-	allErrs = append(allErrs, ValidateNonnegativeField(int64(spec.Replicas), fldPath.Child("replicas"))...)
+	allErrs.AppendWithOrigin("Replicas", ValidateNonnegativeField(int64(spec.Replicas), fldPath.Child("replicas"))...)
```
Wouldn't this want to be "NonNegative" or "Minimum", since that is the actual cause of the error?

yongruilin replied:
How about let to to connect with the DeclarativeValidation tag -- Origin="`minimum`" or "`k8s:minimum`"? Then I think it is better to set the Origin here as 1) here we know we're most likely connecting to the tag `+k8s:minimum` 2) If we want to add the Origin field incrementally, better not to do it inside the `ValidateNonnegativeField()`

thockin replied:
I think aligning origin to openapi makes sense when possible, so "minimum" makes sense to me.

I think we should set it as close to the call to `field.Invalid()` as possible, and if there are wrapper functions which have more meaningful origin (e.g. dns-subdomain vs dns-label) they can replace the origin.

yongruilin replied:
> I think aligning origin to openapi makes sense when possible, so "minimum" makes sense to me.

Sounds good. Updated the code. I think we'd better do a favor for the `MismatchMetric` to let it can easier to compare the Origin from the generated validation and the Origin from the handwritten validation.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7466,7 +7466,7 @@ func validateEndpointAddress(address *core.EndpointAddress, fldPath *field.Path)
 	if address.NodeName != nil {
 		for _, msg := range ValidateNodeName(*address.NodeName, false) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("nodeName"), *address.NodeName, msg))
+			allErrs.AppendWithOrigin("NodeName", field.Invalid(fldPath.Child("nodeName"), *address.NodeName, msg))
```
This is an interesting case.  I think the origin should be the same as the function underlying NodeName (which was given a name alias to hide the details).  Think of what we would put here in the case of a declarative tag, it would be +subfield(name)=+format=dns-label' or something?

I forget which one we use for Node, and am on mobile ATM.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7500,26 +7500,27 @@ func ValidateNonSpecialIP(ipAddress string, fldPath *field.Path) field.ErrorList
 	allErrs := field.ErrorList{}
 	if requireName && len(port.Name) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+		allErrs.Append(field.Required(fldPath.Child("name"), ""))
```
No origin on this?

yongruilin replied:
`NotSupported` could also use it. Since we are adding one more field, would it be better to open the interface for all error types? 
How about replacing the util method as `func (e *Error) WithOrigin(o string) *Error` and `func (list ErrorList) SetOrigin(o string)` so that we can chain the existing error(s) incrementally

thockin replied:
What values might a NotSupported validation set for origin, beyond "not supported" ?

yongruilin replied:
`NotSupported` is not a good example. But isn't that any Error creation interface with a "detail" argument passed would bring in the customized detail string and result in the similar condition happens on `Invalid`? Just trying to understand why need to treat Invalid() special here.

thockin replied:
I'm mostly looking at the pattern of how error Types are used - any given field will have at most on one "Required" or "NotSupported" or "Forbidden" error, but they might have any number of "Invalid" errors.  Setting Origin to "Required" or "Forbidden" seems entirely redundant with the Type field.

Maybe there's use for Origin with other Types, I just have not seen it yet :)

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7500,26 +7500,27 @@ func ValidateNonSpecialIP(ipAddress string, fldPath *field.Path) field.ErrorList
 	}
 	if len(port.Protocol) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("protocol"), ""))
+		allErrs.Append(field.Required(fldPath.Child("protocol"), ""))
```
Origin "Required"?

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7500,26 +7500,27 @@ func ValidateNonSpecialIP(ipAddress string, fldPath *field.Path) field.ErrorList
+		allErrs.Append(field.Required(fldPath.Child("protocol"), ""))
 	} else if !supportedPortProtocols.Has(port.Protocol) {
-		allErrs = append(allErrs, field.NotSupported(fldPath.Child("protocol"), port.Protocol, sets.List(supportedPortProtocols)))
+		allErrs.Append(field.NotSupported(fldPath.Child("protocol"), port.Protocol, sets.List(supportedPortProtocols)))
```
Origin Protocol ?

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7500,26 +7500,27 @@ func ValidateNonSpecialIP(ipAddress string, fldPath *field.Path) field.ErrorList
 	}
 	if port.AppProtocol != nil {
-		allErrs = append(allErrs, ValidateQualifiedName(*port.AppProtocol, fldPath.Child("appProtocol"))...)
+		allErrs.Append(ValidateQualifiedName(*port.AppProtocol, fldPath.Child("appProtocol"))...)
```
Origin?

thockin reviewed (commented):
I am on mobile now and will look more later.

thockin reviewed (commented):
I have a "matcher" in a different PR.  I'm going to check out this branch and see if I can patch it in.

Let's make the usage of this as clear as we can, expecting that the our understanding of it will expand as we do more cases.

Then let's let the tests drive which types we do.  So we should pick a small handful of tests and convert those and only add origin to places needed to fix those tests.

Maybe in the commit comment:

The primary goal of the Origin is for a test which exercises validation logic to ensure that the error it got was caused by the reason it expected, without string-matching on the detail string.

Origin should be set in the most deeply nested validation that can still identify the intention of the error.  

For example, consider this type:

```
type Foo struct {
    Spec FooSpec
}

type FooSpec struct {
    Identity FooIdentity
}

type FooIdentity struct {
    // Name is a DNS1132Label (e.g. "my-name").
    // +required
    Name string

    // FQDN is a DNS1132Subdomain (e.g. "my-name.example.com")
    // +required
    FQDN string
}
```

The hand-written validation for this looks something like:

```
func ValidateFoo(foo *Foo) field.ErrorList {
    var allErrs field.ErrorList 
    allErrs = append(allErrs, validateFooSpec(field.NewPath("spec"), &foo.spec)...)
    return allErrs
}

func ValidateFooSpec(fldPath *field.Path, spec *FooSpec) field.ErrorList {
    var allErrs field.ErrorList 
    allErrs = append(allErrs, validateFooIdentity(fldPath.Child("identity"), &spec.identity)...)
    return allErrs
}

func ValidateFooIdentity(fldPath *field.Path, id *FooIdentity) field.ErrorList {
    var allErrs field.ErrorList
    if id.name == "" {
        allErrs = append(allErrs, field.Required(fldPath.Child("name")))
    } else {
        allErrs = append(allErrs, validateDNS1132Label(fldPath.Child("name"), &id.name)...)
    }
    if id.FQDN == "" {
        allErrs = append(allErrs, field.Required(fldPath.Child("fqdn")))
    } else {
        allErrs = append(allErrs, validateDNS1132Subdomain(fldPath.Child("fqdn"), &id.FQDN)...)
    }
    return allErrs
}
```

We know that `validateDNS1132Subdomain()` calls `validateDNS1132Label()` internally, and that `validateDNS1132Label()` calls `ValidateStringLen()`.  So what should the error origins look like?

The two `field.Required()` calls should have an origin which indicates "field was Required" (or no origin since we already have `Error.Type == Required`).

For the `Name` field, the origin should indicate "DNS1132Label", even if the error was caught by the inner call to `ValidateStringLen()`.  This is the specification of the field, and what tests are verifying.

For the `FQDN` field the origin should indicate "DNS1132Subdomain", even if the error was caught by the inner call to `validateDNS1132Label()` or `ValidateStringLen()`.   Again, this is what tests are testing against.

thockin commented:
https://github.com/kubernetes/kubernetes/pull/130388 builds on this and includes the matcher

pohly commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -16944,6 +16979,16 @@ func TestValidateReplicationController(t *testing.T) {
+			}
+		}
+		if len(expectedOrigins) > 0 {
+			t.Errorf("%s: missing errors with origin: %v", k, expectedOrigins.List())
```
If I were to see a test failure with "missing errors with origin", I wouldn't know what that means.

If I understand the purpose of this PR correctly, then the goal is to replace matching of errors by their details with matching by origin. Shouldn't the test then list all expected errors and their origin and then compare two lists?

As it stands now, it checks that one test case with `expectedOrigins: "Replicas"` has at least one error with that origin. It does not fail if there's more than one.

Speaking of "origin": is that meant to identify exactly one error (i.e. like a unique ID for one particular error) or just a sub-system emitting the error?

If the goal is to replace exact matching, then it can't be just for the sub-system.

thockin replied:
@yongruilin is running with an idea I helped seed, so I'll try to take a stab at answering.

> the goal is to replace matching of errors by their details with matching by origin

Yes

> it checks that one test case with expectedOrigins: "Replicas" has at least one error with that origin. It does not fail if there's more than one.

I think that's by design.  Let me explain why, and you can shoot me down.

The idea for this is (partly) because I wanted to refactor some of the validation logic as part of declarative validation (I know, I am changing two things at once).  In particular, the "DNS label" error produces a very long error string which is not a great experience.  I retooled it to produce two DIFFERENT errors - one for "must start/end with..." and one for "must contain only...", since the characters allowed for those are different.  The current mega-string is embedded in dozens or hundreds of tests.

I started updating them, and then I realized that many tests trigger BOTH errors, but a) it wasn't clear which ones were expected for each testcase, and many of the test loops only handled one error.  I started fixing them and realized it was very tedious and worse, the resulting tests were not actually any less brittle.  There has to be a better way.

So I asked myself: what does a test like `TestValidateResourceSlice()` really want to assert when it is sending bad data for `metadata.name` or metadata.labels` ?  Does it want to do a full modulation of all the edge cases (it doesn't do that today) or does it want to know that the appropriate validation (e.g. `ValidateDNS1123Label()`) was called, but delegate the resposibility for edge-cases to the tests for those functions?  I decided the latter was what it really wants.

So I think we need something more precise than `errs[0].Type == "Invalid"` but less brittle than `errs[0].Detail == "a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')"`.  If Origin reflects the INTENTION, e.g. "this field is a DNS 1123 label", then we can assert that we got at least one error of that origin.  We don't care if it is 1 error or 2 or 5, or what the exact detail string is.  The higher-order test is satisfied.

Origin has another nice property - we can fix tests up to use it one at a time, and it (mostly) doesn't impact other tests.  After converting them, they are almost as precise and far less brittle.  If you buy this, I then offered https://github.com/kubernetes/kubernetes/pull/130388 as a way to make these tests even simpler and more consistent.

Lastly, we need this sort of logic to make converting to declarative validation more tractable -- this allows us to ensure that the errors we get declaratively are actually correlated to the errors we get manually.  As we convert things, we have to be looking closely at tests, which gives a good opportunity and incentive to simplify them. 

> If yes, then the error output should include all information about unexpected errors, not just their origin. It's annoying to have test failures where one has to add log output and rerun tests or fire up a debugger to actually get the information one needs to debug.

With the Matcher you get errors like:

```
--- FAIL: TestValidateTopologySpreadConstraints (0.00s)
    --- FAIL: TestValidateTopologySpreadConstraints/missing_MaxSkew (0.00s)
        validation_test.go:22871: expected an error matching:
            {type="Invalid value", field="field[0].maxSkew", origin="Minimum"}
        validation_test.go:22871: unmatched error:
            {type="Invalid value", field="field[0].maxSkew", value=0, origin="", detail="must be greater than zero"}
```

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go:
```diff
@@ -33,6 +33,13 @@ type Error struct {
 	BadValue interface{}
 	Detail   string
+
+	// Origin indicates the source of this error. It is used in testing to
+	// ensure that the expected errors are returned without matching on detail
+	// strings.
```
```suggestion
	// Origin uniquely identifies the source of an error. It is used in testing to
	// to compare expected errors against actual errors independent on contents
	// of detail strings.
```

?

thockin replied:
"source" is probably the wrong word.  "cause" ?  "reason for"?  "trigger" ?

also "independent on contents of detail strings" is awkward?

jpbetz replied:
:+1: New working LGTM

thockin reviewed (commented):
We need a naming convention for origins.  I guess it should be openapi compatible?  Annoyingly that means lower-bumpy-caps (`minItems`) but the argument to `format` is dash-style-names (`date-time`).

 So do we embrace `ipSloppy` or `ip-sloppy` or `IPSloppy` or `format=ip-sloppy`  or `format(ip-sloppy)` or ...?

Do we recommend any prefixing or anything?  e.g. `mycompany.com/foo` might be different than `foo`?

thockin commented:
We're down to details - does anyone (@pohly?) have fundamental concerns?  I'd like to get this in and try using it.  If it doesn't significantly simplfy tests, we can always back it out.

jpbetz commented:
LGTM, pending outstanding comments are resolved.

Since this is not exposed to users, and is purely a utility to make validation test maintenance easier, I'm in favor of getting this merged and iterating as needed.

yongruilin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7485,38 +7485,38 @@ func ValidateNonSpecialIP(ipAddress string, fldPath *field.Path) field.ErrorList
+		allErrs = append(allErrs, field.Required(fldPath.Child("protocol"), "").WithOrigin("required"))
 	} else if !supportedPortProtocols.Has(port.Protocol) {
-		allErrs = append(allErrs, field.NotSupported(fldPath.Child("protocol"), port.Protocol, sets.List(supportedPortProtocols)))
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("protocol"), port.Protocol, sets.List(supportedPortProtocols)).WithOrigin("port-protocol"))
```
Sounds good to me. I was thinking of putting as "enum" but it doesn't help to improve anything.

yongruilin commented on staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go:
```diff
@@ -96,6 +103,17 @@ func (v *Error) ErrorBody() string {
 }
 
+// SetOrigin sets the Origin for the error.
+func (v *Error) SetOrigin(o string) {
```
Removed.

yongruilin commented:
> We need a naming convention for origins. I guess it should be openapi compatible? Annoyingly that means lower-bumpy-caps (`minItems`) but the argument to `format` is dash-style-names (`date-time`).
> 
> So do we embrace `ipSloppy` or `ip-sloppy` or `IPSloppy` or `format=ip-sloppy` or `format(ip-sloppy)` or ...?
> 
> Do we recommend any prefixing or anything? e.g. `mycompany.com/foo` might be different than `foo`?

@thockin I updated the comment on the Origin field to address the naming convention.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6266,7 +6266,7 @@ func ValidateReplicationControllerSpec(spec, oldSpec *core.ReplicationController
 	allErrs = append(allErrs, ValidateNonnegativeField(int64(spec.MinReadySeconds), fldPath.Child("minReadySeconds"))...)
 	allErrs = append(allErrs, ValidateNonEmptySelector(spec.Selector, fldPath.Child("selector"))...)
-	allErrs = append(allErrs, ValidateNonnegativeField(int64(spec.Replicas), fldPath.Child("replicas"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(spec.Replicas), fldPath.Child("replicas")).WithOrigin("minimum")...)
```
I still feel like this should come from inside ValidateNonnegativeField() but if this is my only comment, we can fix up later

yongruilin replied:
Sure, I think putting it inside would hit a much larger scope of codebase, worth a separate PR.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7485,36 +7485,36 @@ func ValidateNonSpecialIP(ipAddress string, fldPath *field.Path) field.ErrorList
 	}
 	for _, msg := range validation.IsValidPortNum(int(port.Port)) {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("port"), port.Port, msg))
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("port"), port.Port, msg).WithOrigin("format=port-num"))
```
I think `format=` only applies to strings, so this will end up being `portNum` ?

yongruilin replied:
Yea, I was putting it as "portNum" before... I think I changed it wrongly.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7485,36 +7485,36 @@ func ValidateNonSpecialIP(ipAddress string, fldPath *field.Path) field.ErrorList
 	allErrs := field.ErrorList{}
 	if requireName && len(port.Name) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), "").WithOrigin("required"))
```
I don't think we should have origin for Required until we know why

yongruilin replied:
Removed. Relying on on error type should be sufficient.

thockin reviewed (commented):
I'm going to approve and any nits can be followup

/approve
/lgtm

yongruilin commented:
> I'm going to approve and any nits can be followup
> 
> /approve /lgtm

Thanks for approval. I addressed the above nits, could you stamp lgtm again?

thockin commented:
/lgtm

yongruilin commented:
For visibility link to:
- https://github.com/kubernetes/enhancements/issues/5073
